File: fp-in-scala/src/main/scala/fpinscala/answers/parallelism/Par.scala
package fpinscala.answers.parallelism

import java.util.concurrent.*

opaque type Par[A] = ExecutorService => Future[A]

object Par:
  // Exercise 7.2: ここでの並行計算 `Par` の内部表現がどのように定義されているか確認せよ。

  extension [A](pa: Par[A]) def run(s: ExecutorService): Future[A] = pa(s)

  def unit[A](a: A): Par[A] =
    es =>
      UnitFuture(
        a
      ) // `unit` is represented as a function that returns a `UnitFuture`, which is a simple implementation of `Future` that just wraps a constant value. It doesn't use the `ExecutorService` at all. It's always done and can't be cancelled. Its `get` method simply returns the value that we gave it.

  private case class UnitFuture[A](get: A) extends Future[A]:
    def isDone = true
    def get(timeout: Long, units: TimeUnit) = get
    def isCancelled = false
    def cancel(evenIfRunning: Boolean): Boolean = false

  // Exercise 7.1: 2つの並列計算 `Par` の結果を組み合わせるメソッド `map2` はどのようなシグネチャになるか考えよ(実装例は以下のとおり)。

  extension [A](pa: Par[A])
    def map2[B, C](pb: Par[B])(
        f: (A, B) => C
    ): Par[C] = // `map2` doesn't evaluate the call to `f` in a separate logical thread, in accord with our design choice of having `fork` be the sole function in the API for controlling parallelism. We can always do `fork(map2(a,b)(f))` if we want the evaluation of `f` to occur in a separate thread.
      es =>
        val af = pa(es)
        val bf = pb(es)
        UnitFuture(
          f(af.get, bf.get)
        ) // This implementation of `map2` does _not_ respect timeouts. It simply passes the `ExecutorService` on to both `Par` values, waits for the results of the Futures `af` and `bf`, applies `f` to them, and wraps them in a `UnitFuture`. In order to respect timeouts, we'd need a new `Future` implementation that records the amount of time spent evaluating `af`, then subtracts that time from the available time allocated for evaluating `bf`.

  extension [A](pa: Par[A])
    // Exercise 7.3: [java.util.concurrent.Future](https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/util/concurrent/Future.html)のタイムアウトが尊重されるように `map2` を実装せよ。

    def map2Timeouts[B, C](pb: Par[B])(f: (A, B) => C): Par[C] =
      es =>
        new Future[C]:
          private val futureA = pa(es)
          private val futureB = pb(es)
          @volatile private var cache: Option[C] = None

          def isDone = cache.isDefined
          def get() = get(Long.MaxValue, TimeUnit.NANOSECONDS)

          def get(timeout: Long, units: TimeUnit) =
            val timeoutNanos = TimeUnit.NANOSECONDS.convert(timeout, units)
            val started = System.nanoTime
            val a = futureA.get(timeoutNanos, TimeUnit.NANOSECONDS)
            val elapsed = System.nanoTime - started
            val b = futureB.get(timeoutNanos - elapsed, TimeUnit.NANOSECONDS)
            val c = f(a, b)
            cache = Some(c)
            c

          def isCancelled = futureA.isCancelled || futureB.isCancelled
          def cancel(evenIfRunning: Boolean) =
            futureA.cancel(evenIfRunning) || futureB.cancel(evenIfRunning)

  def fork[A](
      a: => Par[A]
  ): Par[A] = // This is the simplest and most natural implementation of `fork`, but there are some problems with it--for one, the outer `Callable` will block waiting for the "inner" task to complete. Since this blocking occupies a thread in our thread pool, or whatever resource backs the `ExecutorService`, this implies that we're losing out on some potential parallelism. Essentially, we're using two threads when one should suffice. This is a symptom of a more serious problem with the implementation, and we will discuss this later in the chapter.
    es => es.submit(new Callable[A] { def call = a(es).get })

  // Exercise 7.8: 関数 `fork` は `fork(x) == x` が常に成り立つことが期待される。[java.util.concurrent.Executors](https://docs.oracle.com/en/java/javase/25/docs/api/java.base/java/util/concurrent/Executors.html)の各種 `ExecutorService` 実装に対して、この関数が常に期待通り振る舞うか検討せよ。

  /*
   * 例えば `Executors.newFixedThreadPool(1)` のようなスレッド数が1の `ExecutorService` を用いると、`fork(x) == x` は成り立たない。
   * これは、`fork` が新しいスレッドを要求するため、スレッド数が1の `ExecutorService` では、`fork(x)` を実行するとデッドロックが発生するからである。
   */

  // Exercise 7.9: この `fork` 実装では任意の固定サイズのスレッドプールでデッドロックが発生しうることを示せ。

  // 例えば、スレッド数が2の `ExecutorService` を用いると、 `fork(fork(fork(x)))` や `fork(map2(fork(x), fork(y)))` でデッドロックが発生する。

  def lazyUnit[A](a: => A): Par[A] = fork(unit(a))

  // Exercise 7.4: 任意の関数 `A => B` を受け取って非同期的に結果を返す関数に変換する関数 `asyncF` を実装せよ。

  def asyncF[A, B](f: A => B): A => Par[B] =
    a => lazyUnit(f(a))

  extension [A](pa: Par[A])
    def map[B](f: A => B): Par[B] =
      pa.map2(unit(()))((a, _) => f(a))

    // Exercise 7.7: `y.map(id) == y` を前提に `y.map(g).map(f) == y.map(f compose g)` が成り立つことを証明せよ。

    /*
     * 関数 `f: B => C`, `g: A => B`, `p: D => C`, `q: A => D` について(多相関数 `map` のparametricityにより)
     * f compose g == p compose q ならば _.map(g).map(f) == _.map(q).map(p) が成り立つことから、
     *    y.map(g).map(f)
     * // p: C => C = id, q: A => C = f compose g とすると _.map(g).map(f) == _.map(id).map(f compose g) なので
     * == y.map(id).map(f compose g)
     * // y.map(id) == y なので
     * == y.map(f compose g)
     *
     * ref. https://github.com/quchen/articles/blob/master/second_functor_law.md
     */

  def sortPar(parList: Par[List[Int]]) =
    parList.map(_.sorted)

  // Exercise 7.5: 関数 `sequence` を実装せよ。

  def sequenceSimple[A](pas: List[Par[A]]): Par[List[A]] =
    pas.foldRight(unit(List.empty[A]))((pa, acc) => pa.map2(acc)(_ :: _))

  // This implementation forks the recursive step off to a new logical thread,
  // making it effectively tail-recursive. However, we are constructing
  // a right-nested parallel program, and we can get better performance by
  // dividing the list in half, and running both halves in parallel.
  // See `sequenceBalanced` below.
  def sequenceRight[A](pas: List[Par[A]]): Par[List[A]] =
    pas match
      case Nil    => unit(Nil)
      case h :: t => h.map2(fork(sequenceRight(t)))(_ :: _)

  // We define `sequenceBalanced` using `IndexedSeq`, which provides an
  // efficient function for splitting the sequence in half.
  def sequenceBalanced[A](pas: IndexedSeq[Par[A]]): Par[IndexedSeq[A]] =
    if pas.isEmpty then unit(IndexedSeq.empty)
    else if pas.size == 1 then pas.head.map(a => IndexedSeq(a))
    else
      val (l, r) = pas.splitAt(pas.size / 2)
      sequenceBalanced(l).map2(sequenceBalanced(r))(_ ++ _)

  def sequence[A](pas: List[Par[A]]): Par[List[A]] =
    sequenceBalanced(pas.toIndexedSeq).map(_.toList)

  def parMap[A, B](ps: List[A])(f: A => B): Par[List[B]] = fork:
    val fbs: List[Par[B]] = ps.map(asyncF(f))
    sequence(fbs)

  // Exercise 7.6: 関数 `parFilter` を実装せよ。

  def parFilter[A](l: List[A])(f: A => Boolean): Par[List[A]] = fork:
    val pars: List[Par[List[A]]] =
      l.map(asyncF(a => if f(a) then List(a) else List()))
    sequence(pars).map(
      _.flatten
    ) // convenience method on `List` for concatenating a list of lists

  def equal[A](e: ExecutorService)(p: Par[A], p2: Par[A]): Boolean =
    p(e).get == p2(e).get

  def delay[A](fa: => Par[A]): Par[A] =
    es => fa(es)

  def choice[A](cond: Par[Boolean])(t: Par[A], f: Par[A]): Par[A] =
    es =>
      if cond.run(es).get then
        t(es) // Notice we are blocking on the result of `cond`.
      else f(es)

  // Exercise 7.11: `choiceN` を実装し、それを用いて `choice` を実装せよ。

  def choiceN[A](n: Par[Int])(choices: List[Par[A]]): Par[A] =
    es =>
      val ind = n.run(es).get % choices.size // Full source files
      choices(ind).run(es)

  def choiceViaChoiceN[A](cond: Par[Boolean])(t: Par[A], f: Par[A]): Par[A] =
    choiceN(cond.map(b => if b then 0 else 1))(List(t, f))

  // Exercise 7.12: `choiceMap` を実装せよ。

  def choiceMap[K, V](key: Par[K])(choices: Map[K, Par[V]]): Par[V] =
    es =>
      val k = key.run(es).get
      choices(k).run(es)

  // Exercise 7.13: `flatMap` を実装し、それを用いて `choice` と `choiceN` を実装せよ。

  /* `chooser` is usually called `flatMap` or `bind`. */
  extension [A](pa: Par[A])
    def flatMap[B](choices: A => Par[B]): Par[B] =
      es =>
        val a = pa.run(es).get
        choices(a).run(es)

  def choiceViaFlatMap[A](p: Par[Boolean])(f: Par[A], t: Par[A]): Par[A] =
    p.flatMap(b => if b then t else f)

  def choiceNViaFlatMap[A](p: Par[Int])(choices: List[Par[A]]): Par[A] =
    p.flatMap(i => choices(i))

  // Exercise 7.14: `join` を実装せよ。
  // また、 `flatMap` を用いて `join` を、 `join` を用いて `flatMap` をそれぞれ実装せよ。

  // see nonblocking implementation in `Nonblocking.scala`
  def join[A](ppa: Par[Par[A]]): Par[A] =
    es => ppa.run(es).get().run(es)

  def joinViaFlatMap[A](ppa: Par[Par[A]]): Par[A] =
    ppa.flatMap(identity)

  extension [A](pa: Par[A])
    def flatMapViaJoin[B](f: A => Par[B]): Par[B] =
      join(pa.map(f))

object Examples:
  import Par.*
  def sum(
      ints: IndexedSeq[Int]
  ): Int = // `IndexedSeq` is a superclass of random-access sequences like `Vector` in the standard library. Unlike lists, these sequences provide an efficient `splitAt` method for dividing them into two parts at a particular index.
    if ints.size <= 1 then
      ints.headOption.getOrElse(
        0
      ) // `headOption` is a method defined on all collections in Scala. We saw this function in chapter 3.
    else
      val (l, r) = ints.splitAt(
        ints.size / 2
      ) // Divide the sequence in half using the `splitAt` function.
      sum(l) + sum(
        r
      ) // Recursively sum both halves and add the results together.

File: fp-in-ocaml/exercises/parallelism/par.ml
(* 第7章 純粋関数型の並列処理

   Scala 版では java.util.concurrent の [ExecutorService] と [Future] を利用するが、
   OCaml の標準ライブラリには対応する機能がないため、[Thread] ベースの簡易な実装
   ([Executor], [Future]) を自前で用意している。

   OCaml 5 の systhreads (Thread) は同一ドメイン内ではランタイムロックを共有するため
   並列には実行されない (並行実行のみ)。ただし本章の目的は並列計算を「記述」する
   API の設計であり、そのセマンティクスは Thread でも表現できる。
   真の並列実行が必要な場合は [Domain] や domainslib を利用する。

   なお Exercise 7.3 (タイムアウトを考慮した map2) は Java の
   [Future.get(timeout, unit)] という機構が前提であり、この簡易実装には
   対応する機能がないため省略している。 *)

(** 計算結果をあとから取得できる入れ物。[java.util.concurrent.Future] の簡易版。 *)
module Future : sig
  type 'a t

  val completed : 'a -> 'a t
  (** 完了済みの Future を作る。Scala 版の [UnitFuture] に相当。 *)

  val get : 'a t -> 'a
  (** 計算が完了するまでブロックして結果を取り出す。 計算が例外で終わっていた場合はその例外を送出する。 *)

  val is_done : 'a t -> bool

  val create : unit -> 'a t
  (** 未完了の Future を作る。[Executor] が利用する内部 API。 *)

  val deliver : 'a t -> ('a, exn) result -> unit
  (** 結果を書き込み、待機中のスレッドを起こす。[Executor] が利用する内部 API。 *)
end = struct
  type 'a state = Pending | Resolved of ('a, exn) result
  type 'a t = { mutex : Mutex.t; cond : Condition.t; mutable state : 'a state }

  let create () =
    { mutex = Mutex.create (); cond = Condition.create (); state = Pending }

  let completed a =
    {
      mutex = Mutex.create ();
      cond = Condition.create ();
      state = Resolved (Ok a);
    }

  let deliver fut result =
    Mutex.lock fut.mutex;
    fut.state <- Resolved result;
    (* [broadcast] で待機中のすべてのスレッドを起こす *)
    Condition.broadcast fut.cond;
    Mutex.unlock fut.mutex

  let get fut =
    Mutex.lock fut.mutex;
    let rec wait () =
      match fut.state with
      (* [Condition.wait] はミューテックスを解放して通知を待ち、 起こされたら再度ロックを取得して戻ってくる *)
      | Pending ->
          Condition.wait fut.cond fut.mutex;
          wait ()
      | Resolved r -> r
    in
    let r = wait () in
    Mutex.unlock fut.mutex;
    match r with Ok a -> a | Error e -> raise e

  let is_done fut =
    Mutex.lock fut.mutex;
    let d = match fut.state with Pending -> false | Resolved _ -> true in
    Mutex.unlock fut.mutex;
    d
end

(** 固定サイズのスレッドプール。[java.util.concurrent.ExecutorService] の簡易版。 *)
module Executor : sig
  type t

  val make : int -> t
  (** 指定した数のワーカースレッドを持つプールを作る *)

  val submit : t -> (unit -> 'a) -> 'a Future.t
  (** タスクをキューに追加し、結果を受け取るための Future を返す *)

  val shutdown : t -> unit
  (** キューに残ったタスクの完了を待ってすべてのワーカーを停止する *)
end = struct
  type t = {
    mutex : Mutex.t;
    cond : Condition.t;
    tasks : (unit -> unit) Queue.t;
    mutable stopped : bool;
    mutable workers : Thread.t list;
  }

  let rec worker_loop pool =
    Mutex.lock pool.mutex;
    while Queue.is_empty pool.tasks && not pool.stopped do
      Condition.wait pool.cond pool.mutex
    done;
    if Queue.is_empty pool.tasks then
      (* 停止指示済みかつタスクなし: ワーカーを終了する *)
      Mutex.unlock pool.mutex
    else begin
      let task = Queue.pop pool.tasks in
      Mutex.unlock pool.mutex;
      task ();
      worker_loop pool
    end

  let make size =
    let pool =
      {
        mutex = Mutex.create ();
        cond = Condition.create ();
        tasks = Queue.create ();
        stopped = false;
        workers = [];
      }
    in
    pool.workers <- List.init size (fun _ -> Thread.create worker_loop pool);
    pool

  let submit pool f =
    let future = Future.create () in
    let task () =
      let result = try Ok (f ()) with e -> Error e in
      Future.deliver future result
    in
    Mutex.lock pool.mutex;
    Queue.push task pool.tasks;
    Condition.signal pool.cond;
    Mutex.unlock pool.mutex;
    future

  let shutdown pool =
    Mutex.lock pool.mutex;
    pool.stopped <- true;
    Condition.broadcast pool.cond;
    Mutex.unlock pool.mutex;
    List.iter Thread.join pool.workers
end

type 'a t = Executor.t -> 'a Future.t
(** 並列計算の記述。[Executor.t] を受け取って [Future.t] を返す関数として表現する (Exercise 7.2 「Par
    の表現を考えよ」の解答に相当)。 *)

(** 並列計算を実際に実行して結果を取り出す *)
let run (es : Executor.t) (pa : 'a t) : 'a = Future.get (pa es)

(** 定数値をそのまま返す並列計算。[Executor] を利用せず、常に完了済みの Future を返す。 *)
let unit (a : 'a) : 'a t = fun _ -> Future.completed a

(* [map2] は [f] の適用自体を別スレッドで評価しない。並列性の制御は [fork] だけが
   担うという設計方針のため。[f] の評価を別スレッドで行いたい場合は
   [fork (lazy (map2 f a b))] とすればよい。 *)
let map2 (f : 'a -> 'b -> 'c) (pa : 'a t) (pb : 'b t) : 'c t =
 fun es ->
  let fa = pa es in
  let fb = pb es in
  Future.completed (f (Future.get fa) (Future.get fb))

(* もっとも素朴な [fork] の実装。ただし外側のタスクが内側のタスクの完了を待って
   ブロックするため、1つの計算にワーカースレッドを2つ消費してしまう。
   固定サイズのプールでは [fork] の入れ子がプールのサイズを超えるとデッドロックする。
   この問題は本章の後半で議論される (Scala 版 Nonblocking.scala を参照)。 *)
let fork (pa : 'a t Lazy.t) : 'a t =
 fun es -> Executor.submit es (fun () -> run es (Lazy.force pa))

(* Scala の名前渡し引数 (=> A) は OCaml では [Lazy.t] で表現する *)
let lazy_unit (a : 'a Lazy.t) : 'a t = fork (lazy (unit (Lazy.force a)))

(** [fork] と異なりワーカースレッドを消費せず、評価だけを [run] まで遅延する *)
let delay (pa : 'a t Lazy.t) : 'a t = fun es -> Lazy.force pa es

(** Exercise 7.4: [lazy_unit] を用いて、任意の関数 [f] を非同期に結果を評価する関数に変換する[async_f]を実装せよ。
*)
let async_f (_f : 'a -> 'b) : 'a -> 'b t = failwith "Not implemented"

let map (f : 'a -> 'b) (pa : 'a t) : 'b t = map2 (fun a () -> f a) pa (unit ())
let sort_par (p : int list t) : int list t = map (List.sort compare) p

(** Exercise 7.5: Par のリストを1つの Par にまとめる[sequence]を実装せよ。

    - [sequence_right]: リストを右からたどり、再帰のステップを [fork] で別スレッドに逃がす実装
    - [sequence_balanced]: 配列を半分に分割して両側を並列に処理する実装
    - [sequence]: [sequence_balanced] を利用したリスト向けのインターフェース *)

let rec sequence_right (_ps : 'a t list) : 'a list t =
  failwith "Not implemented"

let rec sequence_balanced (_ps : 'a t array) : 'a array t =
  failwith "Not implemented"

let sequence (_ps : 'a t list) : 'a list t = failwith "Not implemented"

(* リストの各要素への関数適用を並列に行う。1つの [fork] で包むことで、
   [par_map] の呼び出し自体は即座に返る *)
let par_map (f : 'a -> 'b) (ps : 'a list) : 'b list t =
  fork (lazy (sequence (List.map (async_f f) ps)))

(** Exercise 7.6: リストの要素の絞り込みを並列で行う[par_filter]を実装せよ。 *)
let par_filter (_p : 'a -> bool) (_ps : 'a list) : 'a list t =
  failwith "Not implemented"

(** 2つの並列計算が同じ結果になるかを確認する *)
let equal (es : Executor.t) (p1 : 'a t) (p2 : 'a t) : bool =
  run es p1 = run es p2

(* [cond] の結果を待ってから実行する側の計算を選ぶ *)
let choice (cond : bool t) (if_true : 'a t) (if_false : 'a t) : 'a t =
 fun es -> if run es cond then if_true es else if_false es

(** Exercise 7.11: [n] の結果を添字としてリストから計算を選ぶ[choice_n]を実装せよ。
    また、[choice_n]を用いて[choice_via_choice_n]を実装せよ。 *)

let choice_n (_n : int t) (_choices : 'a t list) : 'a t =
  failwith "Not implemented"

let choice_via_choice_n (_cond : bool t) (_if_true : 'a t) (_if_false : 'a t) :
    'a t =
  failwith "Not implemented"

(** Exercise 7.12: [key] の結果をキーとして連想リストから計算を選ぶ[choice_map]を実装せよ。 (Scala 版の [Map]
    の代わりに連想リストを利用する) *)
let choice_map (_key : 'k t) (_choices : ('k * 'v t) list) : 'v t =
  failwith "Not implemented"

(** Exercise 7.13: [choice], [choice_n], [choice_map] を一般化した[chooser]を実装せよ。
    また、[chooser]を用いて[choice_via_chooser], [choice_n_via_chooser]を実装せよ。 *)

let chooser (_f : 'a -> 'b t) (_pa : 'a t) : 'b t = failwith "Not implemented"

let choice_via_chooser (_cond : bool t) (_if_true : 'a t) (_if_false : 'a t) :
    'a t =
  failwith "Not implemented"

let choice_n_via_chooser (_n : int t) (_choices : 'a t list) : 'a t =
  failwith "Not implemented"

(* [chooser] は一般に flatMap や bind と呼ばれる *)
let flat_map (f : 'a -> 'b t) (pa : 'a t) : 'b t = chooser f pa

(** Exercise 7.14: 入れ子になった並列計算を平坦化する[join]を実装せよ。
    また、[flat_map]を用いた[join_via_flat_map]、[join]を用いた [flat_map_via_join]も実装せよ。 *)

let join (_ppa : 'a t t) : 'a t = failwith "Not implemented"
let join_via_flat_map (_ppa : 'a t t) : 'a t = failwith "Not implemented"

let flat_map_via_join (_f : 'a -> 'b t) (_pa : 'a t) : 'b t =
  failwith "Not implemented"

module Examples = struct
  (* 本章冒頭の例: 分割統治による合計。
     配列 ([Array]) はリストと異なり、半分への分割が効率的に行える。
     この時点ではまだ並列化されていないことに注意。 *)
  let rec sum (ints : int array) : int =
    if Array.length ints <= 1 then if Array.length ints = 0 then 0 else ints.(0)
    else begin
      let mid = Array.length ints / 2 in
      let l = Array.sub ints 0 mid in
      let r = Array.sub ints mid (Array.length ints - mid) in
      sum l + sum r
    end
end
